applications/welcome/models/access_control.py

```diff
@@ -0,0 +1,61 @@
+#########################################################################
+## This file defines access control methods, including verification, setting and deleting.
+## - access_view can view only.
+## - access_setting can change settings, and view variables/output, but cannot change procedure code.
+## - access_manage can do everything the main user can do.
+#########################################################################
+
+
+#########################################################################
+# Verification methods
+#########################################################################
+
+# return if the user has authority to view device
+def access_view(user_id, device_token):
+    return auth.has_permission(name="view", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+# return if the user has authority to configure device
+def access_setting(user_id, device_token):
+    return auth.has_permission(name="setting", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+# return if the user has authority to share the device with others
+def access_manage(user_id, share_user_email, device_token):
+    return auth.has_permission(name="manage", group_id=(device_token + "_" + user_id), user_id=user_id)
+
+
+#########################################################################
+# Setting methods
+#########################################################################
+
+# give user_id "view" access to "device_token"
+def set_access_view(user_id, device_token):
+    return auth.add_permission(name="view", group_id=(device_token + "_" + user_id), user_id=user_id)
+
```

Comment:
Yes, creating our own access table is more suitable here. Originally we thought it is appropriate to use the built-in methods in web2py, but it turns out to be more complex. I will create our own tables next commit.